pkg/vulnsrc/redhat-csaf/advisory.go

```diff
@@ -0,0 +1,216 @@
+	// cf. https://issues.redhat.com/browse/SECDATA-1115
+	if module == "" {
+		if rpmmod := purl.Qualifiers.Map()["rpmmod"]; rpmmod != "" {
+			// rpmmod = "389-ds:1.4:8100020240613122040:25e700aa"
+			// module = "389-ds:1.4"
+			parts := strings.SplitN(rpmmod, ":", 3)
+			if len(parts) >= 2 {
+				module = parts[0] + ":" + parts[1]
+			}
+		}
+	}
+
+	return &Product{
+		Module:  module,
+		Package: purl,
+		Stream:  cpe,
+	}, nil
+}
+
+func (a CSAFAdvisory) LookUpProductReference(productID csaf.ProductID) (*csaf.ProductIdentificationHelper, error) {
+	eb := oops.Tags("product_reference").With("product_reference", productID)
+	productReference, found := a.findProductIdentificationHelper(productID, a.ProductTree)
+	if !found {
+		return nil, eb.Errorf("product reference not found")
+	}
+	return productReference, nil
+}
+
+func (a CSAFAdvisory) LookUpRelatesToProductReference(productID csaf.ProductID) (csaf.CPE, string, error) {
+	eb := oops.Tags("relates_to_product_reference").With("relates_to_product_reference", productID)
+
+	// Check if the package is a module
+	rel := a.LookUpRelationship(productID)
+	if rel != nil {
+		return a.LookUpModule(rel)
+	}
+
+	product, err := a.LookUpProductReference(productID)
+	if err != nil {
+		return "", "", eb.Wrap(err)
+	} else if product.CPE == nil {
+		return "", "", eb.Errorf("stream CPE not found")
+	}
+
+	return *product.CPE, "", nil
+}
+
+// LookUpModule looks up the module information for a given relationship.
+// e.g.
+//
+//	{
+//	  "category": "default_component_of",
+//	  "full_product_name": {
+//	    "name": "httpd:2.4:8070020230131172653:bd1311ed as a component of Red Hat Enterprise Linux AppStream (v. 8)",
+//	    "product_id": "AppStream-8.7.0.Z.MAIN:httpd:2.4:8070020230131172653:bd1311ed"
+//	  },
+//	  "product_reference": "httpd:2.4:8070020230131172653:bd1311ed",
+//	  "relates_to_product_reference": "AppStream-8.7.0.Z.MAIN"
+//	 }
+func (a CSAFAdvisory) LookUpModule(relationship *csaf.Relationship) (csaf.CPE, string, error) {
+	eb := oops.Tags("lookup_module")
+	if lo.FromPtr(relationship.ProductReference) == "" {
+		return "", "", eb.Wrapf(errUnexpectedRecord, "empty product reference")
+	}
+	eb = eb.With("product_reference", *relationship.ProductReference)
+
+	// Look up the module stream
+	moduleRef, err := a.LookUpProductReference(*relationship.ProductReference)
+	if err != nil {
+		return "", "", eb.Wrap(err)
+	} else if moduleRef == nil || moduleRef.PURL == nil {
+		return "", "", eb.Wrapf(errUnexpectedRecord, "module purl not found")
+	}
+	eb = eb.With("module_purl", *moduleRef.PURL)
+
+	purl, err := packageurl.FromString(string(*moduleRef.PURL))
+	if err != nil {
+		return "", "", eb.Wrapf(err, "invalid purl")
+	} else if purl.Type != "rpmmod" { // Must be "rpmmod" for modular streams
+		return "", "", eb.With("purl_type", purl.Type).Wrapf(errUnexpectedRecord, "unexpected purl type")
+	}
```

Comment:
IIUC we shouldn't return error for `rpm` type:
https://github.com/knqyf263/trivy-db/blob/e581cfc764b0a1b874cd30616a991e55aac2b344/pkg/vulnsrc/redhat-csaf/advisory.go#L50-L62

also we can move this code here.